drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+
+    def declare_services(self):
+       '''
+       Conveniently declares all the services offered by the driver
+       '''
+       services = {
+                      'get_pulse_heading'        : GetPulseHeading,
+                      'estimate_pinger_position' : EstimatePingerPosition,
+                      'reset_position_estimate'  : ResetPositionEstimate,
+                      'start_bagging'            : StartBagging,
+                      'save_bag'                 : SaveBag,
+                      'set_frequency'            : SetFrequency
+                  }
+       [rospy.Service('passive_sonar/' + s[0], s[1], getattr(self, s[0])) for s in services.items()]
+
+    def bag_data(self, signals, trans, rot):
+        if self.bagging:
+            if signals.size == None:
+                self.signals_bag = self.signals_bag.reshape(0, signals.shape[0],
+                                                            signals.shape[1])
+            self.signals_bag = np.stack((self.signals_bag, [signals]))
+            self.trans_bag = np.stack((self.trans_bag, [trans]))
+            self.rot_bag = np.stack((self.rot_bag, [rot]))
+
+
+    def get_dtoa(self, signals):
+        '''
+        Returns a list of difference in time of arrival measurements for a signal between
+        each of the non_reference hydrophones and the single reference hydrophone.
+
+        signals - (self.receiver_count x self.signal_size) numpy array. It is assumed that the
+            first row of this array is the reference signal.
+        
+        returns: list of <self.receiver_count> dtoa measurements in units of microseconds.
+        '''
+        sampling_T = 1.0 / self.sampling_frequency
+        upsamp_T = sampling_T / self.upsampling_factor
+        t_max = sampling_T * signals.shape[1]
+
+        t = np.arange(0, t_max, step=sampling_T)
+        t_upsamp = np.arange(0, t_max, step=upsamp_T)
+
+        signals_upsamp = np.array([np.interp(t_upsamp, t, x) for x in signals])
+
+        dtoa, cross_corr, t_corr = \
+            map(np.array, zip(*[get_time_delta(t_upsamp, non_ref, signals_upsamp[0]) for non_ref \
+                                in signals_upsamp[1 : self.receiver_count]]))
+
+        t_corr = t_corr[0]  # should all be the same
+        self.visualize_dsp(t_upsamp, signals_upsamp, t_corr, cross_corr)
+
+        print "dtoa: {}".format(np.array(dtoa)*1E6)
+        return dtoa
+
+    #Passive Sonar Services
+
+    @thread_lock(lock)
+    def get_pulse_heading(self, srv):
+        '''
+        Returns the heading towards an active pinger emmiting at <self.target_frequency>.
+        Heading will be a unit vector in hydrophone_array frame
+        '''
+        success, err_str = True, ''
+        signals, p0, R = None, None, None
+        try:
+            time = rospy.Time.now()
+
+            # Poll input source until it reports it is ready
+            self.input_source.input_request()
+            while not self.input_source.ready:
```

Comment:
`and not rospy.is_shutdown():`
Even though you time out below, I feel like it'd be safer to have the control-c option in case of anomaly.